### DISCUSSION TITLE
feature: Screen Shake

### DISCUSSION THREAD

**Discussion by Erald557** [AscensionGameDev/Intersect-Engine#1551] [Ideas]
**Is your feature request related to a problem? Please describe.**
When you attack a monster or take damage sometimes there isnt much effect to the action and its a bit boring 

**Describe the solution you'd like**
A small tint of screen shake whenever we attack or take damage it will give a nice combat effect, also its a good alert to tell if you are getting attacked.

**Additional context**
Whether the game devs want to add it it should be optional

**Comment by pandinocoder** [comment]
My comments from the open (currently draft) PR #1348

I am not sure I'm on-board with a 4-tier shake intensity system, number or enum my opinion remains the same.

- I feel like the user should have their own intensity slider, similar to volume, that goes from 0 to 10 (0 disabled, 10 is full) -- this would get "multiplied" against the camera shake intensity specified for the given trigger
- Various different camera shake triggers will shake the screen with varying intensity
  - e.g. A sword collision will not shake the camera as much as an earthquake cutscene
- Camera shake should not be tied to animations
  - melee attacks (which are not animations), animation-less events, etc. are all legitimate camera shake triggers
  - what happens when we have 10 different animations/camera shake triggers all playing on the screen, all with camera shake on them? which camera shake takes priority? how do we know that the camera shake trigger should actually impart a camera shake on the viewer? (e.g. an earthquake will cause camera shake for all viewers on the map, but a sword collision will only shake the camera for the attacker and the recipient of the attack)
  
I am not opposed to us adding the system that does the actual shaking of the camera _without_ any ability to trigger it, and then separate PRs for different shake triggers in the event you only want to implement one at first.

Possible camera shake triggers:
- when melee attacks actually hit (misses wouldn't shake)
- when melee attacks hit an entity that is actively blocking (or has some sort of knockback on hit)
- when knockback is applied to a user
- some step in an event (for cutscenes)
- casting spells
- on spell hit (shakes the camera for the recipient)
  - optionally shakes the camera for all entities within a specific range whether or not they were directly impacted by the spell, for example a fire bomb spell could shake the screens for everyone within 5 tiles
- when the person is struck (by anything?)

Again, I don't think shake should be tied directly to animations. It _should_ be tied to what triggers the animations though (and more). We also need logic for when to ignore certain shake triggers.
